test/k8sT/Services.go

```diff
@@ -1548,6 +1549,133 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			// Check if the client pod is able to get a response from the server once it's up and running
+			Expect(res.WaitUntilMatch("client received")).To(BeNil(),
+				"%s is not in the output after timeout", res.GetStdOut())
+			defer func() {
+				cancel()
+				res.WaitUntilFinish()
+			}()
+		})
+
+		AfterFailed(func() {
+			kubectl.CiliumReport("cilium service list", "cilium bpf lb list")
+			kubectl.LogsWithLabel(helpers.DefaultNamespace, serverPodLabel)
+			kubectl.LogsWithLabel(helpers.DefaultNamespace, clientPodLabel)
+		})
+
+		AfterAll(func() {
+			_ = kubectl.Delete(gracefulTermYAML)
+		})
+
+		It("Checks graceful termination on service endpoints deletion", func() {
+			By("Deleting service endpoint")
+			var wg sync.WaitGroup
+			wg.Add(1)
+			// Delete the service pod asynchronously as the following steps need
+			// to be checked while the pod is terminating.
+			go func() {
+				defer wg.Done()
+				res := kubectl.DeleteResource("pod", fmt.Sprintf("-n %s -l %s", helpers.DefaultNamespace, serverPodLabel))
+				ExpectWithOffset(1, res).Should(helpers.CMDSuccess(), "Unable to delete %s pod", serverPodLabel)
+			}()
+
+			By("Waiting until server is terminating")
+			ctx, cancel := context.WithCancel(context.Background())
+			res := kubectl.LogsStream(helpers.DefaultNamespace, serverPod, ctx)
+			Expect(res.WaitUntilMatch("terminating")).To(BeNil(),
+				"%s is not in the output after timeout", res.GetStdOut())
+			defer func() {
+				cancel()
+				res.WaitUntilFinish()
+			}()
+
+			// The client pod exits with status code 0 on graceful termination.
+			By("Checking if client pod terminated successfully")
+			filter := `{.status.phase}`
+			status, err := kubectl.GetPods(helpers.DefaultNamespace, clientPod).Filter(filter)
+			Expect(err).Should(BeNil(), "Failed to get pod status %s", clientPod)
+			Expect(status).Should(BeIdenticalTo("Succeeded"), "Unexpected pod status %q", status)
+			time.Sleep(10 * time.Second)
+
+			By("Checking if terminating service endpoint doesn't serve new connections")
+			res = kubectl.ApplyDefault(testPodYAML)
+			Expect(res).Should(helpers.CMDSuccess(), "Unable to apply %s", testPodYAML)
+			err = kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testDSClient", helpers.HelperTimeout)
+			Expect(err).Should(BeNil(), "Pods %s failed to come up", testPodLabel)
+			pods, err := kubectl.GetPodNames(helpers.DefaultNamespace, testPodLabel)
+			Expect(err).Should(BeNil(), "Cannot retrieve pod names by filter %s", testPodLabel)
+			for _, pod := range pods {
+				// Repeat the command as Kubernetes events propagation may have delays
+				Eventually(func() bool {
+					res = kubectl.ExecPodCmd(
+						helpers.DefaultNamespace, pod,
+						helpers.CurlFail("graceful-term-svc.default.svc.cluster.local.:8081"))
+					return !res.WasSuccessful()
+				}, 2*time.Minute, time.Second).Should(BeTrue(), "%q can connect when it should not work \n", pod)
+			}
+			wg.Wait()
```

Comment:
Could we also assert whether the backend gets eventually removed as expected after graceful period? Meaning, to ensure that agent internal refcounting etc is all still working as expected?

Reply:
That's validated in unit tests as that's the rightful place to map clean-up assertions.